### DISCUSSION TITLE
JWT tokens get too large when users have access to many groups

### DISCUSSION THREAD

**Discussion by hybridherbst** [verdaccio/verdaccio#2698] [Q&A]
As of 5.x, both the UI and API use JTW tokens by default.

It seems that these, currently, contain a full list of groups and real_groups, which means they're growing large very fast.
Are there plans to remedy this? JWT doesn't seem to scale at all even for medium-sized orgs and teams, e.g. having access to ~50 repos seems to already cause the headers to become too big.

Right now, I'm unsure what the right place to remedy this would be; seems JWT tokens are so deeply integrated now that even setting legacy: true will only do _something_ for the API side, not the web interface.

Any thoughts?

Potentially related: #1413, #1530, https://github.com/n4bb12/verdaccio-github-oauth-ui/issues/139

**Comment by juanpicado** [comment]
With the new plugin architecture I would like anyone is able to override core functionality, I have nothing in progress just giving thoughts 💭 deeply on my mind.  I understand the pain for this.

**Reply by hybridherbst** [reply]
On that note: one main goal for JWT tokens seems to be to reduce the number of steps for auth requests (aka: the token already contains the allowed groups). This has some processing upsides but otherwise a lot of UX downsides: e.g. when a user's access rights in the backend change, they'll need to request a new token instead of keeping the existing one and automatically having more access.

Is this really the intended behaviour? I'd argue that "user pain" is always a bigger driver than "small performance overhead", and/or that it would be great if either Verdaccio itself or a plugin could allow the config maintainer to choose between these. There are cases where access configuration should be fully delegated to "somewhere else", e.g. I'd want the config to be minimal (just "*": ask_this_plugin) and then a plugin not only saying yes/no but actually returning a list of packages where the user has access.
Is that possible right now and/or in that new architecture?

**Reply by juanpicado** [reply]
sorry the delay (heavy end of year)

> On that note: one main goal for JWT tokens seems to be to reduce the number of steps for auth requests (aka: the token already contains the allowed groups). This has some processing upsides but otherwise a lot of UX downsides: e.g. when a user's access rights in the backend change, they'll need to request a new token instead of keeping the existing one and automatically having more access.

Yes, that's the whole idea. JWT was introduced as optional if in some cases the trade-off were cheap enough based in your implementation, but again, optional, furthemore, the name legacy also had some downsides that pushed to implement JWT, but since improvement in Node.js and the new Crypto signature https://github.com/verdaccio/verdaccio/pull/1953 for legacy this might change a bit.

> Is this really the intended behaviour? I'd argue that "user pain" is always a bigger driver than "small performance overhead", and/or that it would be great if either Verdaccio itself or a plugin could allow the config maintainer to choose between these. There are cases where access configuration should be fully delegated to "somewhere else", e.g. I'd want the config to be minimal (just "*": ask_this_plugin) and then a plugin not only saying yes/no but actually returning a list of packages where the user has access.
Is that possible right now and/or in that new architecture?

That's the part we need to design. The current behaviour is due few facts, the origins of sinopia design still remains in the current architecture (I personally bet more in stability rather flexibility) but that's the reason why v6 now looks different, I expect this version to be better, flexible and at the same time stable. 

I haven't started with plugins architecture yet, I've been studying, reading, checking other architectures (projects), reading contributors suggestions and requests so we can make it scalable and flexible, so we are in the perfect moment to make big changes (reasonable of course) and of course :) need help. Feel free to drop your ideas whenever you like on the #1690 ticket (listed all plugins ticket attached)
